hazelcast/sql.py

```diff
@@ -0,0 +1,1629 @@
+        so that it can be raised to the user.
+
+        Args:
+            error (Exception): The error to reraise.
+            connection (hazelcast.connection.Connection): Connection
+                that the query requests are routed to. If it is not
+                live, we will inform the user about the possible
+                cluster topology change.
+
+        Returns:
+            HazelcastSqlError: The reraised error.
+        """
+        if not connection.live:
+            return HazelcastSqlError(
+                self.get_client_id(),
+                _SqlErrorCode.CONNECTION_PROBLEM,
+                "Cluster topology changed while a query was executed: Member cannot be reached: %s"
+                % connection.remote_address,
+                error,
+            )
+
+        if isinstance(error, HazelcastSqlError):
+            return error
+
+        return HazelcastSqlError(
+            self.get_client_id(), _SqlErrorCode.GENERIC, try_to_get_error_message(error), error
+        )
+
+    def close(self, connection, query_id):
+        """Closes the remote query cursor.
+
+        Args:
+            connection (hazelcast.connection.Connection): Connection
+                that the first execute request, hence the close request
+                must route to.
+            query_id (_SqlQueryId): The query id to close.
+
+        Returns:
+            Future:
+        """
+        request = sql_close_codec.encode_request(query_id)
+        invocation = Invocation(request, connection=connection)
+        self._invocation_service.invoke(invocation)
+        return invocation.future
+
+
+class SqlExpectedResultType(object):
+    """The expected statement result type."""
+
+    ANY = 0
+    """
+    The statement may produce either rows or an update count.
+    """
+
+    ROWS = 1
+    """
+    The statement must produce rows. An exception is thrown is the statement produces an update count.
+    """
+
+    UPDATE_COUNT = 2
+    """
+    The statement must produce an update count. An exception is thrown is the statement produces rows.
+    """
+
+
+class SqlStatement(object):
+    """Definition of an SQL statement.
+
+    This object is mutable. Properties are read once before the execution
+    is started. Changes to properties do not affect the behavior of already
+    running statements.
+    """
+
+    TIMEOUT_NOT_SET = -1
+
+    TIMEOUT_DISABLED = 0
+
+    DEFAULT_TIMEOUT = TIMEOUT_NOT_SET
+
+    DEFAULT_CURSOR_BUFFER_SIZE = 4096
+
+    def __init__(self, sql):
+        self.sql = sql
+        self._parameters = []
+        self._timeout = SqlStatement.DEFAULT_TIMEOUT
+        self._cursor_buffer_size = SqlStatement.DEFAULT_CURSOR_BUFFER_SIZE
+        self._schema = None
+        self._expected_result_type = SqlExpectedResultType.ANY
+
+    @property
+    def sql(self):
+        """str: The SQL string to be executed."""
+        return self._sql
+
+    @sql.setter
+    def sql(self, sql):
```

Comment:
If we can add api doc for setters I would rather add them. Please add "Raises:" too